Fix undesired UX behavior of DataFrame output in IPython Notebook

danieljl opened the pull request:
Closes #10231

shoyer commented:
Could you please squash this to a single commit and add this to the release notes for 0.16.2? I think this would go in under "Other API changes": https://github.com/pydata/pandas/blob/master/doc/source/whatsnew/v0.16.2.txt

@jorisvandenbossche @jreback any thoughts on this issue? see also the associated issue #10231

danieljl committed:
Fix undesired UX behavior of DataFrame output in IPython Notebook #10231

danieljl commented:
@shoyer I've squashed it into a single commit. Btw, I think this is more appropriate added to release notes under "Other enhancements", isn't it?

shoyer commented on doc/source/whatsnew/v0.16.2.txt:
```diff
@@ -25,6 +25,8 @@ New features
 Other enhancements
 ^^^^^^^^^^^^^^^^^^
 
+- Fix undesired UX behavior of DataFrame output in IPython Notebook. (:issue:`10231`)
```
I might say: "Removed duplicate scroll bars in DataFrame HTML representation in IPython Notebook" (just to be a little more descriptive)

jreback replied:
yeh, this needs a bit more explanation. Possibly even a screen shot. Its *seems* trivial, but people complain about UX changes the most!

shoyer commented:
@danieljl This is a (small) API change, so I would put it under that category. "Other enhancements" is really for fully backwards compatible changes, e.g., adding a new optional keyword argument.

jorisvandenbossche commented:
Personally I don't think it is needed to put it under API changes, as it is not really something that will break people's code, or something for which people have to adjust there code, it is only a visual change. I would keep 'API changes' for things people have to pay attention to if they use that in their code.

jorisvandenbossche commented:
I was only thinking: how does this behave in older versions of IPython? (eg somebody who is still using IPython 2?)

shoyer commented:
@jorisvandenbossche would you call this a bug fix? It's kind of a strange case -- we had a work around, which was resolved in a later version of IPython.

See the linked issue for screenshots.

jorisvandenbossche commented:
Maybe we can detect the IPython version and dependable on that leave the max-width in or not?

@shoyer hmm, maybe also not a bug fix. I would keep it at an 'enhancement' of the visual output of a dataframe in the notebook. But does not really matter that much I think

jreback commented:
I agree with @jorisvandenbossche I would detect ipython version and not make any change for ipython < 3.

Further, I think this needs a separate section in the what's new with a picture. UX changes needs to be announced loudly (even if its is a 'bug' fix).

jreback commented:
@danieljl can you update with ipython version detection?

danieljl commented:
@jreback done. However, in IPython 2 horizontal scroll bar is still unscrollable unless the page is refreshed.

jreback commented on pandas/core/frame.py:
```diff
@@ -22,6 +22,7 @@
 from numpy import nan as NA
 import numpy as np
 import numpy.ma as ma
+import IPython
```
this is not a required dependency, so do this in the actual function. surround with a ``try/except``. if you cannot figure out the version (e.g. ipython is not installed), then use the new style is ok.

jreback commented on pandas/core/frame.py:
```diff
@@ -538,8 +539,12 @@ def _repr_html_(self):
 
-            return ('<div style="max-height:1000px;'
-                    'max-width:1500px;overflow:auto;">\n' +
+            if IPython.version_info < (3, 0, 0):
```
use ``LooseVersion`` to test here.

jreback commented on pandas/core/frame.py:
```diff
@@ -538,8 +538,18 @@ def _repr_html_(self):
+            try:
+                import IPython
+
+                if IPython.version_info < (3, 0, 0):
```
``IPython.version_info < LooseVersion("3.0.0")``

danieljl replied:
@jreback `IPython.version_info` is a tuple. Did you mean `IPython.__version__`? Btw, why is this more preferred?

jreback commented:
Put this code as a method in ``HTMLFormatter``.

This needs to be changed a bit. Instead of simply having the ``div_style`` wrap the ``self.html(...)`` method in ``__repr_html__``, [here](https://github.com/pydata/pandas/blob/master/pandas/core/frame.py#L541), I would pass a new parameter to ``to_html``, call it ``notebook`` (pass ``notebook=True``, with a default value of ``notebook=False``), pass this thru to the ``HTMLFormatter`` class (which sets it as an instance variable).

Then just call this code in ``write_result``.

This will put all of the code in the correct place, and make it obvious what is going on.

danieljl commented:
@jreback Should it be like this 821eaff?

danieljl commented:
@jreback Should it be like this 7e73bec?

danieljl commented:
@jreback Should it be like this 3e0c851?

jreback commented on doc/source/whatsnew/v0.16.2.txt:
```diff
@@ -25,6 +25,8 @@ New features
 Other enhancements
 ^^^^^^^^^^^^^^^^^^
 
+- Removed duplicate scroll bars in DataFrame HTML representation in IPython Notebook. (:issue:`10231`)
```
for notebook >= version 3.0.0

jreback commented:
pls squash this to a single commit

jreback commented on doc/source/whatsnew/v0.16.2.txt:
```diff
@@ -25,6 +25,8 @@ New features
 Other enhancements
 ^^^^^^^^^^^^^^^^^^
 
+- Removed duplicate scroll bars in DataFrame HTML representation for IPython 3.0.0 or later. (:issue:`10231`)
```
This needs some addtl explanation w.r.t. ``max_rows``. IOW, I find that using ``pd.set_setoption('max_rows,None)`` is now a nice way to use the notebook (for a terminal I have this set to a smaller value, however). maybe mention of same in ``options.rst``

jreback commented:
@shoyer @jorisvandenbossche @TomAugspurger @sinhrks 

thoughts?

TomAugspurger commented:
looks good.

sinhrks commented:
lgtm.

shoyer commented:
Yep, looks pretty good to me.

jorisvandenbossche commented:
Yes, here as well. I quickly tested it, and although I was used to having shorter outputs of dataframes in my notebook workflow (now it show about 30 rows, with this change the full 60 rows of a truncated dataframe repr will be visible, leading to a longer output), I think I will get used to the change. 

The only disadvantage is that I find the "toggle output scrolling" feature of IPython (with this, you see around 9 rows) rather ugly (the way too heavy shadow). 
And that I did find the 30 rows you see now a better height as the 60 rows (but that has to do with our default options for max_rows, so another discussion)

danieljl commented:
@jreback Squashed

shoyer commented:
@jorisvandenbossche agreed, it is a little ugly to have a drop shadow here. But this is still better than double scroll bars.

jreback commented:
@danieljl that looks good. can you expand the whatsnew note a bit? approach this from the perspective of someone reading this, and since this is a UX change, we want to be obvious/clear what exactly this is.

danieljl commented:
@jreback Whatsnew updated :)

jorisvandenbossche commented:
One comment: this adds a new `notebook` kwarg to `DataFrame.to_html`, which I don't think is really useful to be exposed to the user? 
Isn't it possible to not add it? (eg leave the logic of that in `__repr__` instead of in `HTMLFormatter`)

danieljl commented:
@jorisvandenbossche Yup, initially I implemented it in `_repr_html_`. But I refactored it as suggested by @jreback in https://github.com/pydata/pandas/pull/10232#issuecomment-110184910.

jreback commented:
I guess we could make it ``_notebook`` e.g. a private argument. This totally belongs in the ``HTMLFormatter`` class (as for example you could then sub-class it and/or use it somewhere else that you are displaying to a notebook)

jorisvandenbossche commented:
OK, agreed. Other option is to call `HTMLFormatter` (or DataFrameFormatter) directly from `__repr__` instead of `to_html`, like

```
return fmt.DataFrameFormatter(self, max_rows=max_rows, max_cols=max_cols,
                              show_dimensions=show_dimensions,
                              notebook=True).to_html()
```

or if you think this is useful in `to_html` as well (and not only in HTMLFormatter), let's just document it in the docstring (and maybe use a more generic name like `div`?)

jreback commented:
how about then we make this an argument to ``to_html`` instead then. (I still would use ``notebook`` as it conveys, hey, we want to output this to a particular medium, html, but needs a special wrapper because its in the notebook)

danieljl commented:
So should the code be refactored again?

jorisvandenbossche commented:
@danieljl I think the latest what @jreback said was to keep it in `to_html`, but then it should get added to the docstring, that is all

jreback commented on pandas/core/frame.py:
```diff
@@ -1387,7 +1387,8 @@ def to_html(self, buf=None, columns=None, col_space=None, colSpace=None,
                                            max_cols=max_cols,
-                                           show_dimensions=show_dimensions)
+                                           show_dimensions=show_dimensions,
+                                           notebook=notebook)
```
here pass directly to ``.to_html()`` (and add in the doc-string). That way this is not relevant to ``DataFrameFormatter``, just to the ``HTMLFormatter`` (which is created in the ``.to_html`` call)

danieljl commented:
If everything's okay, I will squash it again.

jreback commented:
lgtm. go ahead and squash

danieljl commented:
Done

jreback commented:
merged via https://github.com/pydata/pandas/commit/814dbe8e206e3867a5970aaf45f4b7029ad2f59d

thanks!
